text/0000-proc-macros.md

```diff
@@ -0,0 +1,417 @@
+to type information, etc. This would allow some really interesting tools.
+However, it has some large downsides - it complicates the whole compilation
+process (not just the macro system), it pollutes the whole compiler with macro
+knowledge, rather than containing it in the frontend, it complicates the design
+of the interface between the compiler and macro, and (I believe) the use cases
+are better addressed by compiler plug-ins or tools based on the compiler (the
+latter can be written today, the former require more work on an interface to the
+compiler to be practical).
+
+We could have a dedicated syntax for procedural macros, similar to the
+`macro_rules` syntax for macros by example. Since a procedural macro is really
+just a Rust function, I believe using a function is better. I have also not been
+able to come up with (or seen suggestions for) a good alternative syntax. It
+seems reasonable to expect to write Rust macros in Rust (although there is
+nothing stopping a macro author from using FFI and some other language to write
+part or all of a macro).
+
+For attribute-like macros on items, it would be nice if we could skip parsing
+the annotated item until after macro expansion. That would allow for more
+flexible macros, since the input would not be constrained to Rust syntax. However,
+this would require identifying items from tokens, rather than from the AST, which
+would require additional rules on token trees and may not be possible.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+### macros with an extra identifier
+
+We currently allow procedural macros to take an extra ident after the macro name
+and before the arguments, e.g., `foo! bar(...)` where `foo` is the macro name
+and `bar` is the extra identifier. This is used for `macro_rules` and is useful
+for macros which define classes of items, rather than instances of items. E.g.,
+a `struct!` macro might be used similarly to the `struct` keyword.
+
+My feeling is that this macro form is not used enough to justify its existence.
+From a design perspective, it encourages uses of macros for language extension,
+rather than syntactic abstraction. I feel that such macros are at higher risk of
+making programs incomprehensible and of fragmenting the ecosystem).
+
+Therefore, I would like to remove them from the language. Alternatively, they
+could be incorporated into the new design by having another kind of macro
+function:
+
+```
+#[macro_with_ident]
+pub fn foo(&Token, TokenStream, &mut MacroContext) -> TokenStream;
+```
+
+where the first argument is the extra identifier.
+
+
+### Linking model
+
+Currently, procedural macros are dynamically linked with the compiler. This
+prevents the compiler being statically linked, which is sometimes desirable. An
+alternative architecture would have procedural macros compiled as independent
+programs and have them communicate with the compiler via IPC.
```

Comment:
I think we can start by running each plugin's registrar and all expansion in a thread dedicated to that plugin.
This will avoid accidental dependence on thread-local state (which we have some of, most notably the string interner).

If we have a `&mut` context, we should be able to temporarily "transfer ownership" to the expanding thread, without the context even implementing `Sync`, just `Send`.